### PR TITLE
Number component edit form

### LINKED ISSUE #30: Implement number component edit form
Builder edit form and preview for the `number` Form.io component type.

### PR DESCRIPTION
**Pull request by sergei-maertens**
Closes #30

- [x] Add number builder form
    - [x] Suffix
    - [x] Decimal limit
    - [x] Allow negative
    - [x] Validate.min
    - [x] Validate.max
- [x] Add preview + preview stories
- [x] Add component edit
- [x] Add and test validations
- [x] Fix min/max validations & default value being empty strings for empty inputs (see https://github.com/jaredpalmer/formik/issues/1869)

### CODE REVIEW

File: src/registry/number/edit.tsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * Form to configure a Formio 'number' type component.
+ */
+const EditForm: EditFormDefinition<NumberComponentSchema> = () => {
+  const [isKeyManuallySetRef, generatedKey] = useDeriveComponentKey();
+  const {errors} = useFormikContext<NumberComponentSchema>();
+
+  const erroredFields = Object.keys(errors).length
+    ? getErrorNames<NumberComponentSchema>(errors)
```
Comment (CharString) [inline src/registry/number/edit.tsx:35]: How should I read this? It's a function from FormikErrors -> string[]  ... But what do the angle brackets tell me?

File: src/registry/number/edit.tsx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    : [];
+  // TODO: pattern match instead of just string inclusion?
+  // TODO: move into more generically usuable utility when we implement other component
```
Comment (CharString) [inline src/registry/number/edit.tsx:38]: We'll know what we need to be able to express. Defining the types, is a bit like unittests, you ultimately I want them to help me make sense of the code and aid in changing the code with trust, not prevent me from doing so.

And aid can come in the form of pointing out all the locations I need to refactor when changing the type.

Having opaque types, can help with that. So we should be careful about which types should end up in that types package, and which will only prevent us from change... Change is inevitable. All we do should help with change.

File: src/registry/number/edit.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  tooltip: '',
+  showInSummary: true,
+  showInEmail: false,
+  showInPDF: true,
+  hidden: false,
+  clearOnHide: true,
+  isSensitiveData: false,
+  defaultValue: undefined,
+  decimalLimit: undefined,
+  allowNegative: false,
```
Comment (CharString) [inline src/registry/number/edit.tsx:139]: I'm not convinced of  having this *and* min. Making impossible states impossible, should be a goal.

**Review by CharString — CHANGES_REQUESTED** [review]

File: src/components/ComponentConfiguration.stories.tsx [hunk 4]
```diff
@@ -278,3 +281,63 @@ export const Email: StoryObj<typeof Template> = {
...
+    expect(await preview.findByText('Updated preview label'));
+
+    const previewInput = preview.getByLabelText<HTMLInputElement>('Updated preview label');
+    await expect(previewInput).toHaveDisplayValue('');
+    await expect(previewInput.type).toEqual('number');
+
+    // Ensure that the manually entered key is kept instead of derived from the label,
+    // even when key/label components are not mounted.
+    const keyInput = canvas.getByLabelText('Property Name');
+    await fireEvent.change(keyInput, {target: {value: 'customKey'}});
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/components/ComponentConfiguration.stories.tsx:336]: I'll clean up the `await` - it is indeed not required. I remember this being some baby steps with interaction testing :)

If I remember correctly, the `fireEvent` was deliberate because clearing the input and using `userEvent.type` results in a derived key being set from the label because the value of the field is briefly an empty string. This form is essentially simulating a user selecting the current value and typing a replacement value (and I didn't manage to do the latter with testing-library).

File: .storybook/main.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@ const config: StorybookConfig = {
     'storybook-react-intl',
   ],
   docs: {
-    autodocs: true,
+    autodocs: 'tag',
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply .storybook/main.ts:17]: Correct! I much prefer the approach in the SDK with explicit MDX files for docs as I find it produces better usable documentation.

File: src/components/ComponentConfiguration.stories.tsx [hunk 6]
```diff
@@ -191,20 +193,21 @@ export const TextField: StoryObj<typeof Template> = {
     await userEvent.click(preview.getByRole('button', {name: 'Add another'}));
     await expect(preview.getByTestId('input-customKey[0]')).toHaveDisplayValue('');
     // test for the default value inputs -> these don't have accessible labels/names :(
-    const addButtons = await canvas.getAllByRole('button', {name: 'Add another'});
+    const addButtons = canvas.getAllByRole('button', {name: 'Add another'});
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/components/ComponentConfiguration.stories.tsx:196]: Observed locally, but unable to reliably fix it. It does seem to be more of a storybook/testing library problem than implementation problem, so as long as CI/headless browser passes, leaving this for now

File: src/components/ComponentPreview.stories.tsx [hunk 7]
```diff
@@ -143,14 +146,15 @@ export const Email = {
...
+    const canvas = within(canvasElement);
 
     // check that the user-controlled content is visible
     await canvas.findByText('Email preview');
     await canvas.findByText('A preview of the email Formio component');
 
     // check that the input name is set correctly
-    const input = await canvas.getByLabelText('Email preview');
+    const input = canvas.getByLabelText('Email preview');
+    // @ts-ignore
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/components/ComponentPreview.stories.tsx:157]: `component.key` - it does not properly propagate that `args.component` satisfies the AnyComponentSchema and thinks it may be `FallbackSchema` . I tried `expect(someTypePredicate(args.component)).toBe(true)` but that also doesn't help with type narrowing, because TS doesn't realize execution stops on failed exception.

File: src/components/formio/affix.stories.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import Affix from './affix';
+
+export default {
+  title: 'Formio/Utils/Affix',
+  component: Affix,
+  parameters: {
+    modal: {noModal: true},
+  },
+  args: {
+    children: 'm<sup>3</sup><sub>kg</sub>',
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/components/formio/affix.stories.ts:12]: not proven with a test, but when I was manually testing it out, it corrected the missing `</sub>` indeed! This is also documented behaviour, I read something about it in their README.

File: src/components/formio/number.tsx [hunk 9]
```diff
@@ -31,17 +49,49 @@ const NumberField: React.FC<JSX.IntrinsicElements['input'] & NumberProps> = ({
...
+  suffix: string;
+  children: React.ReactNode;
+}
+
+const Wrapper: React.FC<WrapperProps> = ({suffix, children}) => {
+  if (!suffix) return <>{children}</>;
+  return (
+    <div className="input-group">
+      {children}
+      <div className="input-group-append">
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/components/formio/number.tsx:90]: for now the goal is reproducing the markup of the existing form builder, and this is what's produced in the form builder. The SDK has different markup & classnames indeed, but bootstrap/formio don't do BEM.

File: src/registry/number/edit-validation.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import {IntlShape} from 'react-intl';
+import {z} from 'zod';
+
+import {buildCommonSchema} from '@/registry/validation';
+
+// undefined (optional) for unspecified, otherwise a finite numeric value. Note that
+// null would be nicer, but formio's schema does not support null for validate.min,
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/registry/number/edit-validation.ts:7]: I suppose it makes sense for whether an optional argument was provided at all, or was passed but with "no value `null`" instead. But I agree that it is dumb.

File: src/registry/number/edit.tsx [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+ * Form to configure a Formio 'number' type component.
+ */
+const EditForm: EditFormDefinition<NumberComponentSchema> = () => {
+  const [isKeyManuallySetRef, generatedKey] = useDeriveComponentKey();
+  const {errors} = useFormikContext<NumberComponentSchema>();
+
+  const erroredFields = Object.keys(errors).length
+    ? getErrorNames<NumberComponentSchema>(errors)
+    : [];
+  // TODO: pattern match instead of just string inclusion?
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/registry/number/edit.tsx:37]: Copied over code, so this will be refactored at some point.

TS supports template literals and allows you to produce mapped/aliased types, e.g. it is possible to do something like

```
`[key: p as get${p}]: () => {...}`
```
where `p` is the keys of an object to make a readonly accessor.

There are then some crazy generics constructions combined with `keyof` and recursion that *I think* should make it possible to generate a typescript list/union of all possible nested key paths, e.g. `'type' | 'key' | 'validate.min' | ...` and so on, from a given component schema.

So, I think it's possible to get a type safe function where we could then use expressions like `validate.*` or something like that with a combination of regex/explicit names/arrays etc.

So long story short - I'm not entirely sure yet what the best approach here would be, but type safe checks feel like they're within reach.

File: src/registry/number/edit.tsx (hunk 1 above)
Comment (sergei-maertens) [reply src/registry/number/edit.tsx:35]: `useFormikContext<Values>` type var describes the shape/type of `values` (and therefore also the structure of `errors`, but there the type of every value in the key-value pair is `string`)

`getErrorNames` is our own function, not from formik, but is inspired by it and follows the same approach. I think the `getErrorNames` type arg might be possible to remove because it can be inferred from the `error` argument type :thinking:

File: src/registry/number/edit.tsx (hunk 2 above)
Comment (sergei-maertens) [reply src/registry/number/edit.tsx:38]: This TODO was not about the types package, but internally in the builder library!

The intent of the types package is to model the real world:

* strongly type any extensions we add to the formio schemas, which can be relevant for builder, renderer *and even* the backend (e.g. the date `openForms.minDate` configuration in the builder gets transformed by the backend to `datePicker.minDate`, so this requires `openForms.minDate` typing for the builder (and backend) and `datePicker.minDate` typing for the renderer & backend.
* remove the bits from formio that don't apply to our product(s) - this aids developers who look at formio and implement things in preventing them from implementing things that will never be executed. That was something I saw happening a lot when Sven was working on the renderer

It does create some more hurdles - if we implement or change functionality in our backend/frontend, this means that the types package must be updated first to model this, but I don't think it will be that problematic since it's mostly in the "concrete" domain (as opposed to the domain of generics).

File: src/registry/number/edit.tsx (hunk 1 above)
Comment (sergei-maertens) [reply src/registry/number/edit.tsx:35]: Some quick testing - it is indeed not needed, I will finish the date component PR and then do a refactor round!

image
